src/StripePopulator/InvoicePopulatorTrait.php

```diff
@@ -0,0 +1,67 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace ZfrCash\StripePopulator;
+
+use DateTime;
+use ZfrCash\Entity\Invoice;
+use ZfrCash\Entity\LineItem;
+
+/**
+ * @author  Michaël Gallego <mic.gallego@gmail.com>
+ * @licence MIT
+ */
+trait InvoicePopulatorTrait
+{
+    /**
+     * @param Invoice $invoice
+     * @param array   $stripeInvoice
+     */
+    protected function populateInvoiceFromStripeResource(Invoice $invoice, array $stripeInvoice)
+    {
+        $invoice->setCreatedAt((new DateTime())->setTimestamp($stripeInvoice['created']));
+        $invoice->setPeriodStart((new DateTime())->setTimestamp($stripeInvoice['period_start']));
+        $invoice->setPeriodEnd((new DateTime())->setTimestamp($stripeInvoice['period_end']));
+        $invoice->setSubtotal($stripeInvoice['subtotal']);
+        $invoice->setTotal($stripeInvoice['total']);
+        $invoice->setTax(isset($stripeInvoice['tax']) ? $stripeInvoice['tax'] : 0);
+        $invoice->setTaxPercent(isset($stripeInvoice['tax_percent']) ? $stripeInvoice['tax_percent'] : 0);
+        $invoice->setAmountDue($stripeInvoice['amount_due']);
+        $invoice->setCurrency($stripeInvoice['currency']);
+        $invoice->setClosed($stripeInvoice['closed']);
+        $invoice->setPaid($stripeInvoice['paid']);
+        $invoice->setForgiven($stripeInvoice['forgiven']);
+        $invoice->setAttemptCount($stripeInvoice['attempt_count']);
+        $invoice->setDescription($stripeInvoice['description']);
+
+        // Create the latest 10 line items for details. Note that an invoice may have more line items,
+        // but if it contains thousands of items it may quickly break everything
+        $lineItems = [];
+
+        // @TODO: what will happen here if the invoice is updated? Will we have duplicated line items?
```

Comment:
@Ocramius , this part seems broken. Basically, the idea is that once an invoice is created, Stripe sends a webhook that I process to persist the DB locally (for a lot of legal reasons). Whenever it updates on Stripe, Stripe sends new webhooks for the same invoice, but with some updated properties, so I make sure I have everything up to date too.

In Stripe, an invoice has multiple line items. A line item is a "breakdown" of the invoice, to know where the final amount comes from. I have implemented that as a OneToMany association, however I'm concerned that the following code will completely replace the line items each time I receive a webhook for the same invoice.

I have configured orphanRemoval=true on "Invoice" mapping, so I suspect that Doctrine will take care of deleting previous items, and creating new ones, isn't it?

But I'm not entirely satisfied with it.